src/main/java/com/google/sps/servlets/ProjectServlet.java

```diff
@@ -0,0 +1,329 @@
+    }
+    if (!isEmptyParameter(visibility) &&
+        (visibility.toLowerCase().equals(PUBLIC) ||
+         visibility.toLowerCase().equals(PRIVATE))) {
+      projEntity.setProperty("visibility", visibility);
+    }
+
+    String ownersString = request.getParameter("owners");
+    String editorsString = request.getParameter("editors");
+
+    // If creating:
+    //      owners are current User and any people in parameter
+    //      if no one provided, current User is the only owner
+    // if updating:
+    //      owners are current User and any people in parameter
+    //      if no one provided, change nothing
+    if (isCreateMode || !isEmptyParameter(ownersString)) {
+      ArrayList<String> listOwnerEmails = new ArrayList<String>();
+      if (!isEmptyParameter(ownersString)) {
+        listOwnerEmails = parseEmails(ownersString);
+      }
+      listOwnerEmails.add(uEmail);
+      projEntity.setIndexedProperty("owners",
+                                    withDuplicatesRemoved(listOwnerEmails));
+    }
+
+    // If anything provided for editors, overwrite current editors
+    if (!isEmptyParameter(editorsString)) {
+      ArrayList<String> listEditorEmails = parseEmails(editorsString);
+      projEntity.setIndexedProperty("editors",
+                                    withDuplicatesRemoved(listEditorEmails));
+    }
+
+    // Use URL-safe key as project ID
+    // No security risk: only information exposed are kind ("Project")
+    // and name/ID (autogenerated by Datastore)
+    // Key is incomplete until upserted to Datastore
+    if (isCreateMode) {
+      projEntity.setProperty("proj-id",
+                             KeyFactory.keyToString(datastore.put(projEntity)));
+    }
+
+    datastore.put(projEntity);
+    response.sendRedirect("/");
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response)
+      throws IOException {
+    response.setContentType("application/json");
+
+    UserService userService = UserServiceFactory.getUserService();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Must be logged in
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/");
+      return;
+    }
+
+    String uEmail = userService.getCurrentUser().getEmail();
+    String projId = request.getParameter("proj-id");
+
+    // Will be either a single project based on project ID or one or more
+    // projects from a query based on various parameters
+    ArrayList<Entity> projects = new ArrayList<Entity>();
+
+    // Searching for one project with a project ID
+    if (!isEmptyParameter(projId)) {
+      // Project must be public or User must be an owner or editor for private
+      // projects
+      Key projKey = KeyFactory.stringToKey(projId);
+      Entity projEntity = new Entity("Project");
+      try {
+        projEntity = datastore.get(projKey);
+      } catch (Exception e) {
+        response.sendRedirect("/");
```

Comment:
Server side wouldn't handle this but it would be good to print a message to the user on the DOM saying no project with that key exists.

Reply:
Discussed error handling in chat